[Drake] Vendingmachine step4

DrakeYang opened the pull request:
피드백 부탁드립니다!

DrakeYang committed:
재고정보에 음료타입 추가

DrakeYang committed:
재고정보 내부 변수타입 수정: Drink.Type > DrinkInventory.DrinkType

DrakeYang committed:
수정된 재고정보 적용 : 음료추가, 재고정보리턴

DrakeYang committed:
메인함수 이외의 곳에서 작용할 함수들 타입변수로 선언.

DrakeYang committed:
maxDrinkCount 삭제.
isRightDrinkNumber 함수 수정 : 입력인자에 InventoryDetail 추가. 사용자 입력값에 재고개수와 직접비교

DrakeYang committed:
메인함수의 음료구입관련 함수들 자판기 내부로 이동.
자판기 내부에 프로토콜 선언, 메인함수에서 프로토콜 타입으로 자판기 선언.
재고확인,음료구입,음료이동 함수들 쪼개서 세분화.
메인메뉴에 리턴값 문자열 추가. 액션이 성공하면 성공메세지 출력.

DrakeYang committed:
구매된 음료 재고창고>주문목록 으로 이동하는 기능 추가.
구매된 음료 완료메세지에 추가.
재고를 전부 사려고 할때 나는 에러 수정.
메뉴번호와 주문번호 불일치 수정. 메뉴번호-1 을 해야 배열에 맞는데 +1 이였음

DrakeYang committed:
초코우유,콜라 음료객체를 받아서 음료타입으로 리턴하는 함수 추가

DrakeYang committed:
추가된 음료타입 함수 적용

DrakeYang committed:
모든 음료가 떨어질 경우 에러메세지 추가. noDrinks

DrakeYang committed:
모든음료가 떨어질 경우 프로그램 종료처리 추가

DrakeYang committed:
모든재고 소진 메세지 후 종료메세지 나오도록 수정

DrakeYang committed:
종료 로직에 주석 추가

DrakeYang committed:
음료 초기적재 방법 수정 : 객체생성후 여러번 추가 -> 객체를 여러개 추가해서 추가

DrakeYang committed:
음료 선택단계 함수들 리팩토링(조개기).
기존의 selectDrink -> buyDrink 로 변경

DrakeYang committed:
음료객체를 받아서 재고정보로 리턴하는 함수 추가

DrakeYang committed:
음료타입을 받아서 재고에서 해당음료를 pop 하는 함수 추가.
음료타입과 개수를 받아서 제거하는 함수 추가

DrakeYang committed:
사용하지 않는 함수들 삭제.
기존 기능중 사용자/관리자 로 분리가능한 함수들 분리

DrakeYang committed:
프로토콜 분리 : VendingMachineMenu > 사용자,관리자용.
나눌수 있는 사용자/관리자 용 함수 나눔

DrakeYang committed:
관리자용 메인함수, 내부 음료제거함수 추가.
음료 추가는 아직 추가안됨

DrakeYang committed:
모드 선택용 메뉴종류,입력함수,체크함수 추가

DrakeYang committed:
메인메뉴 함수 수정 : 입력인자 프로토콜 > 자판기, 모드선택 메세지 출력

DrakeYang committed:
각 메뉴 선택지의 종료부분 설명 세분화

DrakeYang committed:
유저/관리자 입력단계 세분화

DrakeYang committed:
유저/관리자 함수 분리 및 적용

DrakeYang committed:
음료 타입을 받아서 스스로를 복사해서 추가하는 함수 추가

DrakeYang committed:
관리자모드에서 음료추가 함수 추가.
메인함수에 적용

godrm commented on VendingMachine/VendingMachine/Checker.swift:
```diff
@@ -9,47 +9,48 @@
+        switch input {
+        case "1" : return InputView.UserFirstMenu.insertMoney
+        case "2" : return InputView.UserFirstMenu.selectDrink
+        case "q" : return InputView.UserFirstMenu.quit
```
- 리턴 타입을 `InputView.UserFirstMenu`로 지정하면 `.insertMoney` 나 `.quit`만 써주면 됩니다. 
- 이런 경우 `nil`을 만들기 보다는 차라리 `.none` 처럼 의미있는 값으로 써주는게 더 좋습니다.

DrakeYang replied:
.none 추가해서 적용했습니다.

godrm commented on VendingMachine/VendingMachine/DrinkInventory.swift:
```diff
@@ -108,19 +153,20 @@ class DrinkInventory {
+        case is ChocoMilk : return checkChocoMilkType(chocoMilk: drink as! ChocoMilk)
+        case is Coke : return checkCokeType(coke: drink as! Coke)
+        case is TopCoffee : return checkTopCoffeeType(topCoffee: drink as! TopCoffee)
+        case is EnergyDrink : return CheckNEnergyDrinkType(energyDrink: drink as! EnergyDrink)
```
지금은 음료가 몇 개 안되지만, 여기 있는 checkChocoMilkType() 같은 메소드를 만드는게 좋은 선택이 아닙니다. 
왜냐하면 타입이나 음료 데이터가 추가될때마다 코드를 고쳐야 하기 때문입니다.
새로운 데이터가 추가되도 데이터 구조만 바꾸면 동작하도록 만드는게 가장 좋습니다.

DrakeYang replied:
음료타입을 음료에 추가해서 해당 함수 대체하였습니다.

godrm commented on VendingMachine/VendingMachine/DrinkInventory.swift:
```diff
@@ -108,19 +153,20 @@ class DrinkInventory {
+    
     /// 음료수 객체를 받아서 추가
-    func addInventory(undefinedDrink:Any)->InventoryDetail?{
+    func addInventory(undefinedDrink:Any)->StoredDrinkDetail?{
```
Any 타입을 쓰는건 지양하세요. 항상 좀 더 구체적인 타입을 지정하도록 노력하세요. 
이미 음료수를 표현할 수 있는 수퍼클래스가 있지 않나요?

DrakeYang replied:
Any -> Drink 로 수정했습니다!

godrm commented on VendingMachine/VendingMachine/DrinkInventory.swift:
```diff
@@ -130,17 +176,17 @@ class DrinkInventory {
         // 재고가 있을경우
         if let drink = drinkInventory.first {
-            let inventoryDetail = InventoryDetail(drinkName: drink.getName(), drinkPrice: drink.getPrice(), drinkCount: drinkInventory.count)
+            let inventoryDetail = StoredDrinkDetail(drinkName: drink.getName(), drinkPrice: drink.getPrice(), drinkCount: drinkInventory.count,drinkType: checkDrinkType(drink: drink)!)
```
nil 타입을 돌려주는 메소드가 생기면 이렇게 강제 `!` 언랩핑을 하게 됩니다.
꼭 nil 이 필요한지 고민해보세요. 
그리고 nil을 주면 `!` 보다는 옵셔널 바인딩을 하거나 `??` 처리를 꼭 처리해주세요.

DrakeYang replied:
해당 함수 포함 옵셔널 리턴 함수를 최대한 줄여보았습니다.

godrm commented on VendingMachine/VendingMachine/DrinkInventory.swift:
```diff
@@ -149,16 +195,34 @@ class DrinkInventory {
+        }
+        if let drinkDetail = getDrinkDetail(drinkInventory: energyDrinkInventory) {
+            result.storedDrinksDetail.append(drinkDetail)
+        }
```
여기 반복되는 `if let`  구조는 좋지않습니다.
레벨2에서 계속 반복적으로 연습하는게 if 나 switch 로 구분하는 메소드 호출 구조를 개선하는겁니다. 
객체지향으로 작성하는 연습을 좀 더 해보세요.

DrakeYang replied:
getDrinkDetail 함수는 배열에 음료객체가 있으면 음료의 정보를, 재고가 없으면 nil 을 리턴하는 함수입니다.
음료창고가 6개 있기 때문에 각 창고의 정보를 얻어오기위새 6번 if 문을 썼고, 음료가 없는경우는 정보를 추가하지 않고 있는경우만 추가하는 함수입니다.
각 배열을 객체화 시키거나 6개의 배열을 한 객체로 모은다 하더라도 음료창고가 6개니 정보를 리턴하는 함수를 6번 쓰는건 어쩔수 없지않나요?

DrakeYang replied:
if문 6번 + 추가문 6번 이 좀 문제인것 같아서 InventoryDetail 객체 내부에 음료정보를 받는 함수를 추가했습니다.
해당 함수는 음료정보를 받아서 추가하는데 nil 값이면 무시합니다. 
해당 함수를 이용해서 줄여보았습니다.

godrm commented on VendingMachine/VendingMachine/OutputView.swift:
```diff
@@ -15,74 +15,128 @@ struct OutputView {
+            case .noDrinks : return "판매가능한 음료가 없습니다. \(errorMessage.quitMessage.toString)"
+            case .toModeSelect : return "모드선택으로 이동합니다."
+            }
+        }
```
`toString()->String` 은 Java 형태이고, Swift는 description()을 사용하는 프로토콜이 있죠!

DrakeYang replied:
습관이 자꾸 튀어나오는것 같습니다. 수정했습니다!

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -7,16 +7,31 @@
-
-
-class VendingMachine {
+class VendingMachine : VendingMachineUserMenu, VendingMachineAdminMenu {
```
여기서 VendingMachineUserMenu와 VendingMachineAdminMenu 가 혼재해서 너무 길어지는 것 같습니다.
extension 으로 프로토콜을 나눠서 구현해주세요.

DrakeYang replied:
extension 으로 나누었습니다. 보기좋게 나누어졌습니다!

DrakeYang committed:
각 메뉴 선택시 선택지 외 선택시 nil 리턴하는 부분 수정 > .none 타입 추가.
메뉴 정의부부분에 .none 추가.
체크함수에 리턴값에서 옵셔널 지우고 .none 추가.
체크 후 받는 함수들에서 옵셔널 제거.
메인함수에 적용

DrakeYang committed:
음료객체 분류시 입력인자타입 수정 Any > Drink.
음료객체에 음료타입 변수 선언. 최하위 상속 객체에서는 생성시 타입을 입력받지 않고 내부의 변수를 계산해서 타입을 결정.
음료객체를 받아서 음료타입 리턴하는 함수 삭제. 다른 함수들에 변경점 적용

DrakeYang committed:
음료추가함수 입력인자 음료타입에 .none 대응 추가.
none 의 경우 throw 처리

DrakeYang committed:
에러 내용 출력방법 변경 toString() > description.
재고정보에 음료정보를 입력받아 nil이면 무시 아니면 추가하는 함수 추가.
전체재고정보 출력 함수에 해당 함수 적용

DrakeYang committed:
자판기 선언부분 프로토콜 정리 : 유저,관리자 프로토콜 모두 준수 -> extention 으로 역할별로 나눔

DrakeYang commented:
피드백 부탁드립니다!

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -258,6 +222,43 @@ class VendingMachine : VendingMachineUserMenu, VendingMachineAdminMenu {
+        case .coke : return drinkInventory.getDrinkDetail(drink: drinkInventory.popCokeInventory()!)
+        case .zeroCalorieCoke : return drinkInventory.getDrinkDetail(drink: drinkInventory.popZeroCalorieCokeInventory()!)
+        case .hotTopCoffee : return drinkInventory.getDrinkDetail(drink: drinkInventory.popHotTopCoffeeInventory()!)
+        case .energyDrink : return drinkInventory.getDrinkDetail(drink: drinkInventory.popEnergyDrinkInventory()!)
```
`drinkInventory.popEnergyDrinkInventory()!` 처럼 inventory에서 pop 할 때 항상 있다고 가정하고 `!`를 붙인다면
왜 옵셔널이어야 하는거죠? 옵셔널이면 예외처리가 있거나, 예외처리가 필요없으면 옵셔널일 필요가 없는거 아닐까요?

DrakeYang replied:
옵셔널을 좀더 줄여보았습니다. 위 함수는 리팩토링 과정에서 삭제되었습니다.

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -235,6 +165,40 @@ class VendingMachine : VendingMachineUserMenu, VendingMachineAdminMenu {
+        case .coke : return orderCoke()
+        case .zeroCalorieCoke : return orderZeroCalorieCoke()
+        case .hotTopCoffee : return orderHotTopCoffee()
+        case .energyDrink : return orderEnergyDrink()
```
이런 코드도 `order(.chocoMilk)` 나 `order(drinkType)` 형태가 될 수는 없냐는거죠.
메소드의 매개변수 값을 활용해서 넘기도록 구현해야 음료 타입이 추가되더라도 제한적으로 수정할 수 있지 않을까요?
메소드를 계속해서 추가해야만 하는 구조가 아니라 동작을 일반화하고 값만 바뀌도록 만드는 게 좋다는 겁니다.

DrakeYang replied:
음료 한종류만 담당하는 객체 DrinkSlot 을 추가했고, 음료추가함수를 통일했습니다.

DrakeYang committed:
옵셔널바인딩 ! 제거

DrakeYang committed:
음료 한종류를 담는 객체 음료슬롯 생성.
에러메세지 추가 : 자판기 초기 적재 실패

DrakeYang committed:
전체적인 리팩토링 진행.
음료 한종류만을 담당하는 객체 DrinkSlot 추가.
음료추가,제거,재고정보 출력,임료객체 팝 기능등을 리팩토링.
진행된 리팩토링에 맞게 다른 객체,메인함수 적용

DrakeYang committed:
프로젝트 설정에 음료슬롯 파일 추가됨

DrakeYang committed:
음료번호 선택시 실제 메뉴와 비교하는 기능 추가

DrakeYang commented:
음료 한종류를 저장하는 DrinkSlot 객체를 추가하였고, 음료 추가,제거,추출 등의 함수를 리팩토링해서 전체에 적용해보았습니다.
피드백 부탁드립니다 !

godrm commented:
왜 음료수 클래스 속성이 다 private가 사라진거죠?

DrakeYang commented:
원하는 음료를 추가하려면 해당 음료의 값을 그대로 받아와야되서 퍼블릭으로 바꿨습니다.
private 로 하고 각 변수용 getter 를 선언하는게 나을까요?

godrm commented:
객체지향 프로그래밍에서 getter를 만들거나 public으로 만들기 전에 고민할 건
상위모듈에서 그 객체의 속성값을 어떻게 사용하느냐 고민해봐야합니다. 
값을 가져가서 탐색하고 비교하고 계산하지말고 객체가 그 행위를 하도록 위임하는게 좋습니다. 
단지 특정값을 출력하려고 한다면 출력을 위한 결과값을 만드는 인터페이스를 고려해보세요

DrakeYang committed:
isEnoughDrink 재귀부분 수정

DrakeYang committed:
음료복사 기능제거, 음료정보를 전달하는객체 DrinkInformation 생성.
각 음료 클래스에 대응하는 Information 객체들 생성.
음료,음료정보에 서로 상대방 객체를 받아서 생성하는 생성자 추가.
음료의 변수들 fileprivate 로 수정.
재고정보를 담당하는 구조체 StoredDrinkDetail,InventoryDetail 음료객체 파일로 이동.
재고정보 담당구조체 생성자 수정:값입력 > 음료객체 입력.
음료정보 객체를 이용한 전달방법 전체적으로 적용

DrakeYang commented:
음료객체의 변수를 담아서 이동시켜주는 음료정보 객체를 추가해서 복사하도록 수정했습니다.
음료정보가 등장했으니 음료객체의 pupilic 변수들을 fileprivate 로 할수있었습니다.
피드백 부탁드립니다!

godrm commented:
이번 단계를 오래 붙잡고 있는것 같아서 저도 답답한 마음이네요.
하지만 쉽고 단순한 코드를 작성하는 연습이 필요한 것 같습니다. 
오늘 오후에 만나서 1:1 리뷰를 합시다. 수고많았습니다.

DrakeYang committed:
음료슬롯에 제네릭 적용

DrakeYang committed:
제네릭 적용된 음료슬롯 수정부분 적용.
음료슬롯을 배열로 처리하도록 변경. 스위치 함수 제거

DrakeYang committed:
음료슬롯 제네릭 적용 변경점 수정

DrakeYang committed:
isRightDrinkNumber 함수 비교기능 수정 : 메뉴개수+1 > 메뉴개수 로 수정.
매뉴배열번호가 아닌 메뉴배열개수 면 +1은 필요 없음

DrakeYang committed:
음료슬롯이 받은 음료슬롯전체를 추가하는 함수 추가 : addDrinkSlot

DrakeYang committed:
가격체크를 음료객체에서 하도록 수정.
음료슬롯으로 추가하는 변경부분 수정

DrakeYang committed:
가격체크를 음료객체에서 하도록 수정.
음료슬롯으로 추가하는 변경부분 수정.
리팩토링해서 필요없어진 부분 주석처리

DrakeYang committed:
주석처리된 이전기능들 제거

DrakeYang committed:
재고객체에서 숫자를 입력받아서 해당 번호에 해당하는 음료정보를 리턴하는 함수 추가 : selectDrinkDetail

DrakeYang committed:
isRightDrinkNumber 함수 기능 이전, 주석처리

DrakeYang committed:
주문정보(음료타입,음료개수) 를 가지는 객체 추가 : OrderDetail

DrakeYang committed:
메뉴체크 함수 기능 수정 : 문자열을 메뉴로 -> 메뉴 받아서 체크, 잘못됬으면 에러를 던짐

DrakeYang committed:
문자열 받아서 메뉴로 바꾸는 함수 추가

DrakeYang committed:
음료관련 함수들에 주문정보를 사용하도록 적용

DrakeYang committed:
음료주문 부분에 주문정보 객체 사용하도록 수정.
입,출력 부분 메인함수쪽으로 이동.
음료 정보 체크 및 개수,가격 체크 기능 삭제. 해당 기능들 음료객체와 메인함수로 이동.

DrakeYang committed:
프로젝트파일 수정됨

DrakeYang committed:
음료주문 부분에 주문정보 객체 사용하도록 수정.
자판기객체 안에 있던 인풋뷰,아웃풋뷰 사용부분 메임함수로 이동.
음료선택,음료개수입력 부분 공통함수로 분리.
사용자,관리자 메뉴 에서 잘못된 메뉴 선택세 메인메뉴로 가던 에러 수정.

DrakeYang commented:
피드백 적용하였습니다.
1. 음료 객체안에 자기 복제 함수 넣고 음료정보 객체 삭제
2. 음료 값 비교 함수 음료객체 안으로 이동
3. 음료슬롯에 제네릭 적용, 스위치 케이스 구문 삭제
4. 자판기 내부에서 인풋뷰 부르는 부분 메임함수로 이동

godrm commented:
수고많았습니다. 이번 미션에서 음료 객체와 그걸 다루는 객체, 자판기 로직을 구성하는 객체들 관계를 잘 정리해보세요.
